### PR TITLE
Feature 10 Ability to log in

### PR DESCRIPTION
**Pull request by IvanitskyiVolodymyr**

### CODE REVIEW

File: backend/Recodo.API/Controllers/AuthController.cs [hunk 1]
```diff
@@ -35,8 +40,19 @@ public async Task<ActionResult<AuthUserDTO>> Register([FromBody] NewUserDTO user
                 User = createdUser 
             };
 
-            return new JsonResult(result);
+            return Ok(result);
+        }
+
+        [HttpPost]
+        [Route("api/Login")]
```
Comment (Anatolii-Ivanov) [inline backend/Recodo.API/Controllers/AuthController.cs]: You can write 
[HttpPost("api/Login")]

Comment (Anatolii-Ivanov) [reply backend/Recodo.API/Controllers/AuthController.cs]: /api shouldn't be part of routes, you can add it in the startup

File: backend/Recodo.API/Startup.cs [hunk 2]
```diff
@@ -56,6 +56,11 @@ public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
                 app.UseSwaggerUI(c => c.SwaggerEndpoint("/swagger/v1/swagger.json", "Recodo.API v1"));
             }
 
+            app.UseCors(x => x
+            .AllowAnyOrigin()
+            .AllowAnyMethod()
+            .AllowAnyHeader());
```
Comment (Anatolii-Ivanov) [inline backend/Recodo.API/Startup.cs]: One line or tabs

File: frontend/src/app/app-routing.module.ts [hunk 3]
```diff
@@ -3,7 +3,7 @@ import { RouterModule, Routes } from '@angular/router';
 
 const routes: Routes = [
   { path: '', loadChildren: () => import('./modules/workspace/workspace.module').then(m => m.WorkspaceModule) },
-  { path: 'register', loadChildren: () => import('./modules/auth/auth.module').then(m => m.AuthModule) }
+  { path: 'recodo', loadChildren: () => import('./modules/auth/auth.module').then(m => m.AuthModule) }
```
Comment (Anatolii-Ivanov) [inline frontend/src/app/app-routing.module.ts]: Can you make login -> /login, register -> /register, workspace -> /personal or what we have there

File: frontend/src/app/modules/auth/auth-routing.module.ts [hunk 4]
```diff
@@ -1,13 +1,16 @@
...
 
 
 const routes: Routes = [
     {
-        path: '',
-        component: RegisterPageComponent
+        path: 'register', component: RegisterPageComponent
+    },
+    {
+        path: 'login', component:LoginPageComponent
```
Comment (Anatolii-Ivanov) [inline frontend/src/app/modules/auth/auth-routing.module.ts]: Space

File: frontend/src/app/modules/auth/login-page/login-page.component.ts [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+        updateOn: 'change',
+      }],
+      password: [, {
+        validators: [
+          Validators.required,
+          Validators.minLength(8),
+          Validators.pattern('^(?=.*[0-9])(?=.*[a-zA-Z])([a-zA-Z0-9]+)$'),
+        ],
+        updateOn: 'change'
+      }],
```
Comment (Anatolii-Ivanov) [inline frontend/src/app/modules/auth/login-page/login-page.component.ts:48]: Move it to ngOnInit

File: frontend/src/app/modules/auth/auth-routing.module.ts [hunk 6]
```diff
@@ -1,18 +1,24 @@
...
     {
-        path: '',
-        component: RegisterPageComponent
+        path : 'register', component : RegisterPageComponent
+    },
+    {
+        path : 'login', component : LoginPageComponent
+    },
+    {
+        path : '', redirectTo: 'register', pathMatch: 'full'
```
Comment (Anatolii-Ivanov) [inline frontend/src/app/modules/auth/auth-routing.module.ts]: Redirect to login
And match should be none (So, /ajsdgh will be also redirected to login

File: frontend/src/app/services/login.service.ts [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+import { AuthUserDto } from '../models/auth/auth-user-dto';
+import { UserLoginDto } from '../models/auth/user-login-dto';
+import { TokenDto } from '../models/token/token-dto';
+import { UserDto } from '../models/user/user-dto';
+import { ResourceService } from './resource.service';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class LoginService extends ResourceService<UserLoginDto> {
```
Comment (Anatolii-Ivanov) [inline frontend/src/app/services/login.service.ts:13]: At the end should be merged with registerService into AuthService

**Review by Anatolii-Ivanov — DISMISSED** [review]

**Review by VitalikMatvieiev — DISMISSED** [review]

File: backend/Recodo.API/Controllers/AuthController.cs [hunk 8]
```diff
@@ -21,10 +21,14 @@ public AuthController(AuthService authService, UserService userService)
             _userService = userService;
         }
 
-        [HttpPost]
-        [Route("api/Register")]
+        [HttpPost("api/Register")]
```
Comment (Jyuart) [inline backend/Recodo.API/Controllers/AuthController.cs]: Move 'api' to the controller level
```
[ApiController]
[Route("api")]
```

File: backend/Recodo.API/Controllers/AuthController.cs [hunk 9]
```diff
@@ -21,10 +21,14 @@ public AuthController(AuthService authService, UserService userService)
...
         public async Task<ActionResult<AuthUserDTO>> Register([FromBody] NewUserDTO userDTO)
         {
+            if (!ModelState.IsValid)
+            {
+                return BadRequest(modelState: ModelState);
+            }
+
```
Comment (Jyuart) [inline backend/Recodo.API/Controllers/AuthController.cs]: We don't need this. We have fluent validation

File: backend/Recodo.API/Controllers/AuthController.cs [hunk 10]
```diff
@@ -35,8 +39,18 @@ public async Task<ActionResult<AuthUserDTO>> Register([FromBody] NewUserDTO user
...
-            return new JsonResult(result);
+            return Ok(result);
+        }
+
+        [HttpPost("api/Login")]
+        public async Task<ActionResult<AuthUserDTO>> Login([FromBody] LoginUserDTO userDTO)
+        {
+            if (!ModelState.IsValid)
+            {
+                return BadRequest(modelState: ModelState);
```
Comment (Jyuart) [inline backend/Recodo.API/Controllers/AuthController.cs]: The same

File: frontend/src/app/models/auth/auth-user-dto.ts [hunk 11]
```diff
@@ -0,0 +1,7 @@
+import { TokenDto } from '../token/token-dto';
+import { UserDto } from '../user/user-dto';
+
+export interface AuthUserDto{
```
Comment (Jyuart) [inline frontend/src/app/models/auth/auth-user-dto.ts]: Do `ng lint` before committing to be able you do not have any warning
```suggestion
export interface AuthUserDto {
```

File: frontend/src/app/modules/auth/auth-routing.module.ts (hunk 6 above)
Comment (Jyuart) [inline frontend/src/app/modules/auth/auth-routing.module.ts]: I think @Anatolii-Ivanov meant `login`, not `register`

File: frontend/src/app/modules/auth/login-page/login-page.component.ts [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+    });
```
Comment (Jyuart) [inline frontend/src/app/modules/auth/login-page/login-page.component.ts:49]: Put this in a separate method and call it inside `ngOnInit`

File: frontend/src/app/services/login.service.ts [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+
+  private user:UserDto | undefined= {} as UserDto;
```
Comment (Jyuart) [inline frontend/src/app/services/login.service.ts]: Why do you need `undefined` here?

**Review by Anatolii-Ivanov — DISMISSED** [review]

File: frontend/src/app/services/login.service.ts [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+    return observable.pipe(
+        map((resp) => {
+            this._setTokens(resp.body?.token);
+            this.user = resp.body?.user;
+            return resp.body?.user;
+        })
+    );
+  }
+
+  private _setTokens(tokens: TokenDto | undefined) {
```
Comment (Jyuart) [inline frontend/src/app/services/login.service.ts]: Please don't use underscores at the beginning of the names of methods

**Review by Jyuart — CHANGES_REQUESTED** [review]

**Review by Anatolii-Ivanov — APPROVED** [review]

**Review by NickRost — APPROVED** [review]

**Review by Jyuart — APPROVED** [review]

**Pull request merged**